drivers/hmis/lib/hmis_data_cleanup/util.rb

```diff
@@ -27,43 +37,73 @@ def self.clear_enrollment_export_ids!
-    # We could `scope.find_in_batches` and then enrollment.assign_attributes and finally "import" the batch with a conflict key of id.
     def self.assign_missing_household_ids!
-      scope = Hmis::Hud::Enrollment.hmis.where(household_id: nil)
-      Rails.logger.info "Assigning household id to #{scope.size} enrollments"
-      scope.each do |enrollment|
-        hh_id = Hmis::Hud::Base.generate_uuid
-        without_papertrail_or_timestamps do
-          enrollment.update_columns(household_id: hh_id) # skips callbacks
-        end
+      Hmis::Hud::Enrollment.hmis.where(household_id: [nil, '']).find_in_batches do |batch|
+        batch.each { |enrollment| enrollment.HouseholdID = Hmis::Hud::Base.generate_uuid }
+
+        result = Hmis::Hud::Enrollment.import(
+          batch,
+          validate: false,
+          timestamps: false,
+          on_duplicate_key_update: { conflict_target: [:id], columns: [:HouseholdID] },
+        )
+        raise "Failed: #{result.failed_instances}" if result.failed_instances.present?
+
+        Rails.logger.info "Assigned HouseholdID to #{result.ids.count} Enrollments"
       end
     end
 
     # Find any single-member households that do not have a HoH, and make that person the HoH
     def self.make_sole_member_hoh!
-      single_member_households = Hmis::Hud::Enrollment.hmis.
-        group(:household_id).
-        having(nf('COUNT', [:HouseholdID]).eq(1)).
-        pluck(:household_id)
-
-      num = Hmis::Hud::Enrollment.hmis.
-        where(household_id: single_member_households).
-        where.not(relationship_to_hoh: 1).size
-
-      Rails.logger.info "Assigning HoH to #{num} single-member households"
-
       without_papertrail_or_timestamps do
-        Hmis::Hud::Enrollment.hmis.
-          where(household_id: single_member_households).
+        rows_affected = Hmis::Hud::Enrollment.hmis.
+          group(:household_id).
+          having(nf('COUNT', [:HouseholdID]).eq(1)).
           where.not(relationship_to_hoh: 1).
           update_all(relationship_to_hoh: 1) # skips callbacks
+
+        Rails.logger.info "Set HoH in #{rows_affected} single-member households"
+      end
+    end
+
+    # UNTESTED!!!
+    def self.fix_bad_personal_id_references_on_services!
+      data_source_id = GrdaWarehouse::DataSource.hmis.first.id
+      # Find and fix Services where the PersonalID does not match the PersonalID on the referenced Enrollment
+      # We may want to expand this to other Enrollment-related record types
+      GrdaWarehouse::Hud::Service.where(data_source_id: data_source_id).left_outer_joins(:enrollment).
+        where(GrdaWarehouse::Hud::Enrollment.arel_table[:id].eq(nil)).
+        find_in_batches do |batch|
+        # map {EnrollmentID=>PersonalID} for each enrollment referenced by this batch of services
+        eid_to_pid = GrdaWarehouse::Hud::Enrollment.where(
+          data_source_id: data_source_id,
+          EnrollmentID: batch.pluck(:EnrollmentID),
```

Comment:
batch is an array here- pluck works but it's a bit confusing imo
```suggestion
          EnrollmentID: batch.map(&:EnrollmentID),
```